lib/Twig/Extensions/Node/Trans.php

```diff
@@ -149,7 +175,136 @@ protected function compileString(Twig_Node $body)
+     * also needed to mantain the integrity of multiline strings and the
+     * indentation of the source file changed. Take for example:
+     *
+     *  Before (A):
+     *
+     *      {% trans %}
+     *          This is a translatable string
+     *          spanning multiple lines
+     *      {% endtrans %}
+     *
+     *  After (B):
+     *
+     *      {% if some_condition %}
+     *          {% trans %}
+     *              This is a translatable string
+     *              spanning multiple lines
+     *          {% endtrans %}
+     *      {% endif %}
+     *
+     * If the above example, we needed to add some condition and thus added
+     * indentation because of the "if". Then, the translatable string changes
+     * and to Gettext it is a different string because of the different spaces
+     * inside the sentence.
+     *
+     * @param string $msg  The message string to be normalized.
+     * @param string $glue The character used to replace spaces.
+     *
+     * @return string The normalized string.
+     */
+    protected function normalizeString($msg, $glue = ' ')
+    {
+        return preg_replace('/\s+/u', $glue, trim($msg));
+    }
+
+    /**
+     * Guesses a name from a Twig_Node_Expression. First, we try to get the
+     * propossed name from a 'as("name")' filter. If there is no 'as' filter,
+     * try to guess a name from the expression itself.
+     *
+     * @param Twig_Node $expr Expression to guess name from.
+     *
+     * @return string Proposed name.
+     */
+    protected function guessNameFromExpression(\Twig_Node $expr)
+    {
+        // Traverse the Expression AST tree trying to guess a name from 'as'
+        // filters, until we reach a non-filter expression.
+        while ($expr instanceof Twig_Node_Expression_Filter) {
+            // OK, we are still inside a filter, let's look if we got a 'as'.
+            if ($expr instanceof Twig_Extensions_Filter_As) {
+                // 'as' filters always have to have at least one argument and
+                // it is a Constant Expression, guarranteed by the parser.
+                return $expr->getNode('arguments')->getNode(0)->getAttribute('value');
+            }
+            $expr = $expr->getNode('node');
+        }
+
+        // We are now at the AST tree for the main expression. We'll try to
+        // mangle the name from the rest of this nodes.
+        return implode('_', $this->extractNames($expr));
+    }
+
+    /**
+     * Guesses a name from a series of Twig_Expression nodes. If the expression
+     * is too complex to guess a name, it throws a Twig_Error_Syntax.
+     *
+     * @param Twig_Node $node The starting node to extract names from.
+     *
+     * @return array Array of names guessed up to this node.
+     *
+     * @throws Twig_Error_Syntax
+     */
+    protected function extractNames(\Twig_Node $node)
+    {
+        switch (get_class($node)) {
```

Comment:
you should use `instanceof` statements rather than checking `get_class`, to deal better with child classes or changes related to class aliases (here, you specify both names, but such code written earlier with only the non-namespaced names would have broken with the change done in twig reversing the alias and class name)